File: go/day1/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	fmt.Println("AoC Day 1 Part 2")

	if len(os.Args) < 2 {
		fmt.Println("Need to provide input")
		return
	}

	input := os.Args[1]
	processInput(input)
}

func processInput(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	top_three_cals := [3]int{0, 0, 0}
	tmp_cal := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		input := scanner.Text()
		if len(input) < 1 {
			pushPop(&top_three_cals, tmp_cal)
			tmp_cal = 0
		} else {
			cal, err := strconv.Atoi(input)
			if err != nil {
				break
			}
			tmp_cal = tmp_cal + cal
			continue
		}
	}
	pushPop(&top_three_cals, tmp_cal)

	fmt.Println(top_three_cals[0] + top_three_cals[1] + top_three_cals[2])
}

func pushPop(stack *[3]int, value int) {
	if stack[0] < value {
		stack[2] = stack[1]
		stack[1] = stack[0]
		stack[0] = value
	} else if stack[1] < value {
		stack[2] = stack[1]
		stack[1] = value
	} else if stack[2] < value {
		stack[2] = value
	}
}

File: rust/day1/src/bin/part2.rs
use std::env;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

fn main() {
    println!("AoC Day 1 Part 2");
    let args: Vec<String> = env::args().collect();

    if args.len() < 2 {
        println!("Need to provide input");
        return;
    }

    let filename = args[1].clone();
    process_input(filename).expect("Failed to process input");
}

fn process_input(filename: String) -> std::io::Result<()> {
    let file = File::open(filename);
    let file = match file {
        Ok(file) => file,
        Err(error) => return Err(error),
    };

    let buf_reader = BufReader::new(file);

    let mut highest_cal: [u64; 3] = [0, 0, 0];
    let mut tmp_cal = 0;

    for line in buf_reader.lines() {
        let line = match line {
            Ok(line) => line,
            Err(_e) => break,
        };

        if line.len() == 0 {
            push_pop(&mut highest_cal, tmp_cal);
            tmp_cal = 0;
        } else {
            let cal: u64 = line.parse().unwrap();
            tmp_cal = tmp_cal + cal;
        }
    }
    push_pop(&mut highest_cal, tmp_cal);

    println!("{:?}", highest_cal[0] + highest_cal[1] + highest_cal[2]);
    Ok(())
}

fn push_pop(stack: &mut [u64; 3], value: u64) {
    if stack[0] < value {
        stack[2] = stack[1];
        stack[1] = stack[0];
        stack[0] = value;
    } else if stack[1] < value {
        stack[2] = stack[1];
        stack[1] = value;
    } else if stack[2] < value {
        stack[2] = value;
    }
}
